### ISSUE TITLE
An event system for Jupyter

### ISSUE THREAD

**Issue by afshin** [jupyter-server/jupyter_server#780]
*This is a draft document, please feel free to comment and help.*

We have (at least) two concurrent efforts that overlap but are not full implementations of a generic event system for Jupyter in themselves: [`jupyter-telemetry`](https://github.com/jupyter/telemetry) and [`jupyterlab-notifications`](https://github.com/jupytercalpoly/jupyterlab-notifications).

A synthesis of these extensions with *generic* endpoints (i.e., not specifically designed and named for telemetry or notifications) would yield a flexible general-purpose event bus for `jupyter-server`-based applications.

cc: @andrii-i @3coins

## Architecture

The `jupyter-server` package will have some REST endpoints and a WebSocket endpoint to handle event creation and subscription.

The `jupyterlab` and `notebook` packages will have front-end extensions that give users a notification user interface for a subset of the events that come through the server.

#### REST endpoints for CR~~U~~D
- `POST /api/events`: Creates a new event in the system, returns an `HTTP 204` with the newly created event's ID or possibly the reified full event
- `GET /api/events`: The list/query endpoint to retrieve multiple events simultaneously. This may return a list of IDs or fully reified events (perhaps one or the other, perhaps supporting both based on query)
- `GET /api/events/{id}`: Retrieves a specific event, fully reified
- `DELETE /api/events/{id}`: Allows removal of an event by a client. **This needs discussion**

#### WebSocket endpoint for subscription
- `wss://.../api/events/stream` The WebSocket connection that pushes events that are relevant to the subscriber (based on knowledge of the logged-in user and potentially other critera)

### JupyterLab 4 extension
- A `Token` (e.g., `INotifications` or `IEvents`) that exposes an `IDataConnector` for event CRUD and an `ISignal` for event subscription
- A visual UI for an event notification center

### Jupyter Notebook 7 extension
- The `Token` from the JupyterLab extension
- A version of the JupyterLab UI for notifications

**Comment by 3coins** [comment]
@afshin 
Should we just add this to the server or need a new server extension package? Is anyone assigned to this task?

**Comment by afshin** [comment]
@3coins, this should be in the core server.

Currently, no one is specifically assigned. I'd like to see the user interface portion of this landing in JupyterLab and I am happy to work on any part of the stack that helps get us there.

I *think* that the work on this already done in the telemetry space might be farther along than the server extension from the notifications extension, so grafting those handlers into `jupyter-server` might be the best way of bringing this into core.

What are you thinking? Let's have a conversation about this with all the people who have interest and bandwidth to work on it.

**Comment by afshin** [comment]
I think Zach is rounding up interested folks (including you) for a conversation.

We are targeting `jupyter-server` v2 and `jupyterlab` v4

**Comment by rahul26goyal** [comment]
@afshin  / @Zsailer : 
please include me on any meeting that might happen related to this. I am interested to learn more on this area and contribute anyway I can.

**Comment by 3coins** [comment]
As discussed in the server meeting on 05/05/2022, here is an initial list of tasks for the event notification system. This list is by no means final, feel free to add comments or feedback.

1. **Event Bus** - [#820](https://github.com/jupyter-server/jupyter_server/pull/820)
    - A central event bus to relay events 
    - `/api/events/subscribe` - Websocket for subscribing to events
    - A default handler for consuming events

1. **Rest API Endpoints**
    - POST `/api/events` - Rest api to create new events
    - GET `/api/events/schemas` - Rest api to query/list registered schemas (Optional)

1. **Event buffer**
   - A queue/buffer to store undelivered event messages

1. **JupyterLab 4 Event Client (jupyterlab-events)**
    - Reuse jupyterlab-telemetry repo, either rename or copy to jupyterlab-events
    - Remove server endpoints, any redundant server code
    - Update client handlers to use the rest api endpoints 
    - Add websocket handler to enable subscribtion to events

1. **Add Default events in server**
    - Add default events e.g., content handler, kernel events in jupyter server

1. **JupyterLab 4 Updates**
    - Add `jupyterlab-events` as dependency inside JupyterLab
    - Subscribe to default events

1. **Event Notification UI (JupyterLab)**
    - UI updates for event notification 

1. **Jupyter Notebook 7 Updates**
    - Add `jupyterlab-events` as dependency
    - Subscribe to default events
    - Can we reuse event notification UI from JupyterLab?

**Comment by afshin** [comment]
Here is a document we can collaboratively edit so that the front-matter of this issue can have a canonical version that we edit once it is ready: https://hackmd.io/q4Rkq2BaS1SIXvyzt8j1yA

**Comment by Zsailer** [comment]
Thanks for bringing this up, @davidbrochart! I think we're going to see this question/conversation come up multiple times moving forward as we continue pushing Jupyter Server forward, while trying to bring jupyverse to the front.

Let me start by saying—technically, the event system _is_ backend agnostic. We just defined a REST + websocket API for posting/subscribing to events. These are schema/protocol driven. Jupyverse can/should create an implementation of this API. I don't think there is anything tied specifically to Tornado here. Any server implementation will always have to write some server-library-specific code to make it work. Consider, if we started this in jupyverse, how would we port it to jupyter_server? We would have to re-implement the handlers in Tornado and drop the FastAPI specific logic.

That said, under the hood, we depend on jupyter_telemetry (hopefully, switching to `jupyter_events` soon) and you are correct—jupyter_telemetry/events depends on traitlets. 

That's because we needed the Event System API to be configurable. I don't see a way around using `traitlets` for this without switching to some other backwards compatible, backend-agnostic, config-based library. For example, it looks to me that jupyverse/FPS is implementing its own (non-backend agnostic) configuration system, `fps.config`. While I believe FPS offers a much cleaner way to handle config, it's not backwards compatible with Jupyter Server. This might be a place we can improve.

Unfortunately, at this time, I don't see a single solution that would work for both. And while I see jupyverse as our future (it's awesome!), I don't think we should block jupyter_server from making advancements using the older dependencies at this point in time. 

Do you have ideas how to reconcile this?

**Comment by Zsailer** [comment]
"backwards compatible, backend-agnostic, config-based library"

To me, this is the "holy grail".

We could probably get pretty close by 
1. writing logic that translates traitlets config into a pydantic `BaseModel`. 
2. handling traits/fields that "observe" other traits/fields.

**Comment by davidbrochart** [comment]
I meant something more simple, like [this GET handler](https://github.com/jupyter-server/jupyter_server/blob/3e64fa5eef7fba9f8e17c30cec688254adf913bd/jupyter_server/services/contents/handlers.py#L91) calls [this get method](https://github.com/jupyter-server/jupyter_server/blob/3e64fa5eef7fba9f8e17c30cec688254adf913bd/jupyter_server/services/contents/manager.py#L387). If we can have the logic in the get method in a separate package, that's a great step towards backend agnosticism.

**Comment by afshin** [comment]
@jweill-aws the "case study" above is about notifications and the idea is that it becomes an extension's job to manage its state. In the case of notifications, the extension will write events it cares about from the event bus into a SQL database and it will be the job of the client to call `DELETE` to remove those items from the database (*i.e.*, make them "read").

**Comment by Zsailer** [comment]
In https://github.com/jupyter/jupyter_events/pull/2, we've have been discussing the handling of sensitive data in the event system. I'm confident that these are already "solved problems" in other systems, so I need some help gathering information about how to properly do it here. 

In https://github.com/jupyter/jupyter_events/pull/2, I added a required field to every schema, "redactionPolices", that is used to describe the sensitivity of every event property. The event logger can be configured to redact sensitive policies from all data in all events. This data is redacted _before_ the event is ever emitted. This provides a simple way to ensure that sensitive data is never persisted.

On the other hand, if a client (e.g. JupyterLab) builds features that depend on the event system, and these features depend on receiving all of the data, redacted events/data breaks these features. This makes the event system unusable to these features when launching in a data-conscious (i.e. most) environments.

To make the event system useful, we need to a secure way to handle sensitive data in transit, specifically when moving between Jupyter Server and its clients. Today, the event bus added in #820 shuttles unencrypted events to the client across the websocket. Any authenticated websocket client can connect to this websocket and "see" all event data—this obviously isn't a secure approach.

This is where I need some help. What are some known patterns for handling sensitive data in transit from server to client? If we encrypt the data in the server, how do we secure decrypt it in something like JupyterLab?

**Comment by Zsailer** [comment]
The basic "plumbing" for Jupyter server's event system landed here: https://github.com/jupyter-server/jupyter_server/pull/862

We've started logging some events from the contents here: https://github.com/jupyter-server/jupyter_server/pull/954
